### PR TITLE
fix(index): skip not changed chunk files in watch mode (fix #133)

### LINKED ISSUE #133: cache does not work in watch mode
We are using code splitting in dev (watch) mode in order to avoid redownloading everything on page refresh but only changed bundles. But compression plugin keeps generating new compressed artifacts even for not changed bundles. I've tried to enable caching and it did not have any effect.

### PR DESCRIPTION
**Pull request by akosyakov**
This PR contains a:

- [x] **bugfix**
- [ ] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

See https://github.com/webpack-contrib/compression-webpack-plugin/issues/133#issuecomment-535353489

### Additional Info

I have not looked at tests yet, but would like to know whether it is right approach first? cc @evilebottnawi

I've followed this guidelines to check whether a chunk is changed: https://webpack.js.org/contribute/plugin-patterns/#changed-chunks

### CODE REVIEW

**Comment by akosyakov** [comment]
I've fixed tests.

**Comment by akosyakov** [comment]
Published it as `@typefox/compression-webpack-plugin@3.0.0` for downstream testing.

**Comment by akosyakov** [comment]
I'm not sure about the approach. It seems to be alternatively we can base it on https://webpack.js.org/contribute/plugin-patterns/#monitoring-the-watch-graph, but I end up with more involving changes to accomplish it, i.e. I have to watch generated assets and in order to collect paths of generated assets traversing of chunks is required.

**Comment by akosyakov** [comment]
I've refactored a bit to include files from modules as well and added test in watch + development modes.

File: test/__snapshots__/watch.test.js.snap [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  ],
+  Array [
+    "96621d3c37d96ad3bf792fcc848d912f.svg",
+    672,
+  ],
+  Array [
+    "async.js",
+    396,
+  ],
+  Array [
```
Comment (akosyakov) [inline test/__snapshots__/watch.test.js.snap]: here you can see that only gz file for async module is regenerated, since it was changed

**Comment by evilebottnawi** [comment]
It is not good solution because our output is not deterministic and potential break other plugins, we have cache option, why don't use this?

**Comment by akosyakov** [comment]
> we have cache option, why don't use this?

New gzip files with very the same content but different timestamp are emitted with the cache option and breaks express.js resource caching. Is there a way to preserve timestamps with the cache option?

**Comment by akosyakov** [comment]
> It is not good solution because our output is not deterministic and potential break other plugins

I don't think it changes anything without watch mode. For single compilation it works as before. Also it's implemented with the approach suggested by webpack docs https://webpack.js.org/contribute/plugin-patterns/#changed-chunks.

**Comment by evilebottnawi** [comment]
> New gzip files with very the same content but different timestamp are emitted with the cache option and breaks express.js resource caching. Is there a way to preserve timestamps with the cache option?

Can you create reproducible test repo?

**Comment by akosyakov** [comment]
> Can you create reproducible test repo?

@evilebottnawi there is a reference to repo on the issue: https://github.com/webpack-contrib/compression-webpack-plugin/issues/133#issuecomment-535353489 It's configured for watching with cache enabled.

**Comment by evilebottnawi** [comment]
Thanks, ping me tomorrow, i will look at this

**Comment by akosyakov** [comment]
@evilebottnawi ping

**Comment by ernestostifano** [comment]
@evilebottnawi Hi guys! I got here because I was about to open an issue for the same reason @akosyakov did.

I've not fully got through all the changes in this PR, but it seems a little over-complicated. Especially because Webpack already has its own cache/emission logic which could be re-used by simply changing current used hook.

**The solution:**

Currently, "compression-webpack-plugin" is taping into `compiler.hooks.emit` hook. Which is called before the emission process has even started. Just after that, Webpack decides which assets should be emitted, **using its internal cache logic**.

Finally, check-passing assets are effectively written and `compiler.hooks.assetEmitted` hook is called for each one of them. Obviously, this way, compressed assets are generated for every existent asset, even if they will not be effectively emitted at the end. (And worse, during development, all assets are compressed again, even those that has not been changed and already exist in memory filesystem).

Initially I though that `compiler.hooks.emit` hook was being used because at some point Webpack removes the source from emitted assets. However, when `compiler.hooks.assetEmitted` hook is called, source (will be of `Buffer` type) is passed as second argument.

Additionally, this is already done asynchronously using "neo-async" library, so no additional looping would be needed, thus increasing performance.

I think that solution should be clear from here and it should be way more simple than expected. I would be glad to try fixing this issue if you like. Please let me know.

***Note:*** Hope I'm not missing something.

**Comment by akosyakov** [comment]
@ernestostifano if there will be official better solution i'm all in for using it :) So far we were using `@theia/compression-webpack-plugin` for Theia and all downstream projects in dev mode and it did not cause any apparent issues.

**Comment by evilebottnawi** [comment]
@akosyakov sorry for delay, WIP on the problem

**Comment by evilebottnawi** [comment]
Sorry for big delay, fixed in master

**Comment by akosyakov** [comment]
@evilebottnawi thank you, that's fine :)

**Pull request closed without merging**